File: NeetCode/Binary_Search/Search_Insert_Position/search__insert__position.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        if target < nums[0]:
            return 0
        if target > nums[-1]:
            return len(nums)

        left, right = 0, len(nums) - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                return mid
            if nums[mid] > target:
                if mid > 0 and nums[mid - 1] < target:
                    return mid
                else:
                    right = mid - 1
            else:
                left = mid + 1
        return mid


if __name__ == "__main__":
    sol = Solution()
    nums = [-1, 0, 2, 4, 6, 8]
    target = 5
    print(sol.searchInsert(nums, target))
    print("Running Solution...")

File: NeetCode/Binary_Search/Search_Insert_Position/Search_Insert_Position.cpp
// Search_Insert_Position.cpp
//
// Compile locally with:
// clang++ Search_Insert_Position.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int searchInsert(vector<int> &nums, int target)
    {
        if (nums[0] < target)
        {
            return 0;
        }
        if (nums.back() > target)
        {
            return nums.size();
        }
        int left(0), mid(0);
        int right = nums.size() - 1;
        while (left <= right)
        {
            mid = (left + right) / 2;
            if (nums[mid] == target)
            {
                return mid;
            }
            if (nums[mid] < target)
            {
                left = mid + 1;
            }
            else
            {
                right = mid - 1;
            }
        }
        return left;
    }
};

int main()
{
    cout << "Running Search_Insert_Position.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
